### PR TITLE
API: add nvim_win_close() to close window by id

### PR DESCRIPTION
**Pull request by bfredl**
This is for plugin closing a float or other window it has handle to. It is complementary to #9663 which is also user-lever commands.

TODO: needs test with ordinary window, and also with `force`.

### CODE REVIEW

File: src/nvim/globals.h [hunk 1]
```diff
@@ -979,7 +979,9 @@ EXTERN char_u e_notopen[] INIT(= N_("E484: Can't open file %s"));
 EXTERN char_u e_notopen_2[] INIT(= N_("E484: Can't open file %s: %s"));
 EXTERN char_u e_notread[] INIT(= N_("E485: Can't read file %s"));
 EXTERN char_u e_nowrtmsg[] INIT(= N_(
-        "E37: No write since last change (add ! to override)"));
+    "E37: No write since last change (add ! to override)"));
+EXTERN char_u e_nowrtmsg_api[] INIT(= N_(
+    "E37: No write since last change (use force=true to override)"));
```
Comment (justinmk) [inline src/nvim/globals.h]: I don't think this is a good idea. Maintaining different flavors of error messages for the API is extra work for little gain. Alternative is to document that the `force` parameter on the API method means "!" or bang (could rename the param as `bang` as well). It's completely reasonable to expect API users to understand this.

Also adding a `api` parameter and codepath to `ex_win_close` is similarly unpleasant.

Comment (bfredl) [reply src/nvim/globals.h]: Ok, I will revert it.

File: src/nvim/api/window.c [hunk 2]
```diff
@@ -57,7 +58,6 @@ void nvim_win_set_buf(Window window, Buffer buffer, Error *err)
                   window);
   }
 
-  try_start();
   int result = do_buffer(DOBUF_GOTO, DOBUF_FIRST, FORWARD, buf->b_fnum, 0);
   if (!try_end(err) && result == FAIL) {
```
Comment (justinmk) [inline src/nvim/api/window.c]: `try_start` was removed. Is the `try_end` paired with some other `try_start` ? We should try to keep them at the same "level" ..

Comment (bfredl) [reply src/nvim/api/window.c]: no, it was just a mistake.

File: src/nvim/api/window.c [hunk 3]
```diff
@@ -470,3 +471,29 @@ void nvim_win_config(Window window, Integer width, Integer height,
...
+/// Close a window.
+///
+/// This is equivalent to |:close| with count except that it takes a window id.
+///
+/// @param window   Window handle
+/// @param force    Behave like `:close!` The last window of a buffer with
+///                 unwritten changes can be closed. The buffer will become
+///                 hidden, even if 'hidden' is not set.
+
+/// @param[out] err Error details, if any
```
Comment (justinmk) [inline src/nvim/api/window.c]: I think the docstring needs to be "connected" for doxygen to map it. so line 483 should have `///`

**Comment by bfredl** [comment]
WTF doesn't gcc support `(void)try_leave(&tstate, err)` to ignore should-use value? What would `(void)` cast mean other than that?

**Comment by justinmk** [comment]
That's weird. Maybe assign it to `vim_ignored` from globals.h

**Comment by hkupty** [comment]
I just want to hightlight the benefit that this single API function brings:
```lua
  if vim.api.nvim_win_close ~= nil then
    local window = vim.api.nvim_call_function("win_getid", {winnr})
    vim.api.nvim_win_close(window, true)
  else
    local save_winnr = nvim.nvim_call_function("bufwinnr", {nvim.nvim_get_current_buf()})
    vim.api.nvim_command(winnr .. ' wincmd w | q')
    if save_winnr ~= -1 and save_winnr ~= winnr then
      vim.api.nvim_command(save_winnr .. ' wincmd w')
    end
  end
```

Thanks!

**Pull request merged**